Plugins/GitUIPluginInterfaces/ViewModels/GitRepoViewModel.cs

```diff
@@ -0,0 +1,32 @@
+﻿using System.ComponentModel;
+using System.ComponentModel.Design;
+using CommunityToolkit.Mvvm.ComponentModel;
+using CommunityToolkit.Mvvm.Messaging;
+using GitUIPluginInterfaces;
+
+namespace Plugins.GitUIPluginInterfaces.ViewModels;
+
+public interface IGitRepoViewModel : INotifyPropertyChanged
+{
+    IEnumerable<GitRevision> Revisions { get; }
+}
```

Comment:
I'm having troubles connecting the name of the interface and the API it provides. A "model" encapsulate the app's data (i.e., "state"), whereas an interface is a behavioural contract.
I feel this should be called something like `INotifyRevisionsChanged`. ...or better yet `INotifyRepositoryStatusChanged` since to find out whether the repo changed one would typically execute `git status`.

I can't see how `Revisions` property get populated, `GitRepoViewModel` doesn't implement this property... In the current form it feels wrong.

Reply:
We also have an existing mechanism of notifying when repository state changes:
https://github.com/gitextensions/gitextensions/blob/7cee1b824b3fef6a0235a40127e7870fcf7b4b0e/GitUI/GitUICommands.cs#L35
https://github.com/gitextensions/gitextensions/blob/7cee1b824b3fef6a0235a40127e7870fcf7b4b0e/GitUI/GitUICommands.cs#L46-L47

This change feels like we're adding a new notification mechanism. Could you please elaborate on the plan here?

Reply:
Hm, RepoChangedNotifier  is for opening etc, this is for actions currently to refresh the grid. Not sure if the same method can be used?

Reply:
> We also have an existing mechanism of notifying when repository state changes:
> This change feels like we're adding a new notification mechanism. Could you please elaborate on the plan here?

Yes, all this tightly-coupled stuff shall be replaced with:

> a message bus (IMessenger, IRecipient<TMessage>)

step by step as more functionality is moved to the `GitRepoViewModel` (and to additional dedicated ViewModels).
As you mentioned yourself a few weeks ago: When the repo has changed, the necessary information shall be refreshed in a ViewModel.
GUI sections subscribe to the `OnChange` (and `OnChanging`) notifications of the ViewModel and update accordingly using the ViewModel's properties.
Then FormBrowse / RevisionGrid do not need to notify the Left Panel any longer. And all UI elements, particularly the Left Panel, do not need to run git commands.

This PR is the initial step where the ScriptRunner sends the `RepositoryModifiedMessage` to all (ViewModels) which need to react - without knowing them nor with the need to call callbacks.
This resolves the problem that the former implementation needs a script host control instance on which the callback is called.

>> The decoupling by means of the RepoModifiedMessage facilitates the implementation https://github.com/gitextensions/gitextensions/pull/11273.

>> I have prepared MVVM for the imlementation of the GitOutputControl.
>> I am going to use MVVM for the implementation of the FileStatusList using TreeView.

---

> A "model" encapsulate the app's data (i.e., "state"), whereas an interface is a behavioural contract.

A ViewModel provides properties and regarding `OnChange` (and `OnChanging`) events.
An interface with these properties describes this kind of contract (semantics included).

> I feel this should be called something like INotifyRevisionsChanged. ...or better yet INotifyRepositoryStatusChanged since to find out whether the repo changed one would typically execute git status.

That's much too specific. The GitRepoViewModel will have many more properties in the future.
And "git log" and "git status"  shall be run by the ViewModel(s) or an underlying Model but not by the subscribers of the `OnChange` event. The subscribers shall just display the property values.
After running a repo-changing command, we just send the `RepositoryModifiedMessage`.

> I can't see how Revisions property get populated, GitRepoViewModel doesn't implement this property...

At the moment, we just need and implement the notification part. Currently, the property shall not be read. Though its name can be used for the notification event.